Andy dev

andykb9b13 opened the pull request:
I finally finished refactoring the signIn and login forms. They are located in /client/src/pages. 

Here are the highlights of the changes:

- Instead of using constant state changes to keep track of inputs in the form, the form state is now being tracked by react-hook-form. 
- Inputs are "registered" with react-hook-form for it to keep track of them.
` <Input placeholder="John" {...register("firstName")} />`

- Submission of the data is handled through react-hook-form and validation is handled through yup as a resolver in the useForm() method.
``const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm({ resolver: yupResolver(signInSchema) });

- Validation through yup is in a custom hook. 
`const loginSchema = yup.object().shape({
    email: yup.string().required("Please Enter Your Email"),
    password: yup.string().required("Please Enter Your Password"),
  });`

- The signUp and login objects are destructured when the hook is called.
- ` const { signInSchema } = useSchemas();`

mykltronn commented on client/src/hooks/useSchemas.js:
Nice custom hook! Keeps the messy business of validation logic out of the login form, which is generally nice.

I'm curious why you put this much effort in to validating data on the front-end. Does this all get validated on the backend?

andykb9b13 replied:
It had not occurred to me until you wrote this that I should be able to do all of this on the backend as well. Upon investigating Mongoose some more it looks like I should be able to do all of these specific validations in the backend. Honestly, I haven't spent as much time with the Mongoose documentation and for some reason it feels a little challenging to parse sometimes. I'll dive into it more. 

If I think about it, maybe the benefits of doing client-side validation are that data doesn't even get a chance to hit your database unless it's valid which maybe prevents miscellaneous wrong data from hanging out there. If I do client-side AND server-side validation I can see that posing potential problems with keeping track of the shape of the data (i.e. something is required on server-side but not client-side). Is there a case for doing both or one in particular? Are there specific benefits to server-side vs. client-side validation?

mykltronn replied:
So a few things to note here:

You should never rely on front-end validation alone. It's trivial for somebody to bypass validation on the front-end. And, as you say, you sure don't want to define it both places. But your schema is already defined on the back end. When the structure of the data changes, it'll be on the backend. So just keep it all in the same place, I say!

There's this concept in software engineering-- what should this piece of code "know" about? I would say your backend should "know" about the data, the shape of that data, etc. Your frontend should know about style, appearance, user interaction. It should pass anything data related to the backend without any opinion on that data. I think you'll thank yourself for it in the future. This is part of the general concept known as "separation of concerns". 

The only benefit I can think of to validating on the front-end is to skip the request to the backend for invalid data. In my experience, this is very much not worth it.

mykltronn commented on client/src/pages/Login.jsx:
```diff
@@ -1,141 +1,79 @@
+      const { data } = await login({
+        variables: { ...userInput },
       });
+      Auth.login(data);
+      alert("Successfully logged in!");
```
So does `login` here always return the happy path? What if the credentials are wrong? I gather that there's where the `error` object from the request comes in, but what do we pass to `Auth.login` in that case?

andykb9b13 replied:
If the wrong credentials are passed in it will hit the resolver in the server and throw a new error. However, it doesn't look like I am passing that error back to the front-end to inform the user. It is showing up in the console. I'll have to work on that.

mykltronn commented on client/src/pages/Login.jsx:
```diff
@@ -1,141 +1,79 @@
-        <FormLabel htmlFor="password">Password</FormLabel>
+    <Card sx={styles.card}>
+      <Typography level="h2">Login</Typography>
+      <form onSubmit={handleSubmit(onSubmit)}>
```
Is this really the pattern here? react-hook-forms provides `handleSubmit`, to which you must pass a function that actually handles submit? What, then, is it doing? Just validating with the schema stuff?

andykb9b13 replied:
This is indeed the pattern in the docs for react-hook-form. `handleSubmit()` is a wrapper that has additional functionality and options for handling the data. It adds the e.preventDefault() ....by default...🤷‍♂️ (sorry that feels like a bad pun). The second argument in handleSubmit is `onError` which give access to some validation errors. There are some other methods it has that I am not accessing so it may be the case that this is overkill here.

mykltronn commented on client/src/pages/Login.jsx:
```diff
@@ -1,141 +1,79 @@
-    const { name, value } = event.target;
-    setFormData((prevData) => ({ ...prevData, [name]: value }));
-  };
+const styles = {
```
Just a matter of preference, but this seems like it could live outside this module. Maybe in an adjacent style-dedicated file?

andykb9b13 replied:
Agreed. At some point I'd like to pass a theme to the entire front-end. Styling is what I have focused on the least so far in this project but I definitely think having larger bits of styling code in a separate file is where I will go with it.

mykltronn commented on client/src/pages/SignUp.jsx:
```diff
@@ -34,265 +34,75 @@ const styles = {
-      Auth.login(data);
+      Auth.login(data); //signing in the user once created
       alert("Account created!");
     } catch (err) {
```
Obviously you're not tooled to report exceptions to some aggregator like Sentry, but there are two things I'll say about this bit for posterity:
1.This is a classic "swallowed" error. `catch (err)` here will gobble up ANY exception that occurs inside the try. That could be a syntax error, that could be a network connection error, that could be anything.
2. To avoid swallowing this error here and losing the helpful information it might contain forever, you might be tempted to `console.log(err)` or `SomeErrorReportingTool.report(err)`. But that's potentially problematic as well. Does err contain sensitive data? Could it ever? Do we _know_ what data would be considered sensitive here?

This is all to say-- `try-catch` can be a little bit nasty in Javascript in my view.

andykb9b13 replied:
I definitely see that it is going to gobble up any error but I'm not quite sure what to do as an alternative. Any suggestions? I know in python you can target specific exceptions but I haven't seen anything like that in JavaScript up to this point.

mykltronn commented on package.json:
A note on NPM dependencies and `package.json`:
I know it's a common convention to use that caret `^`  when declaring a dependency. As you know, the caret pins the dependency to the major version you've declared. This means that, in the case of `yup` for instance, the package author might make releases all the way up to `1.12.3` and your package would be updated with it.

In production settings, this is a nightmare! A more common convention is to pin to the patch version, which is accomplished using the tilde `~` instead of the caret. But, to be honest, I prefer to pin to the exact version-- major minor and patch. Here that would be accomplished by just saying `"yup": "1.2.0".

Some reading on [semver here](https://semver.org/) and[ some reading on npm dependency version pinning](https://docs.npmjs.com/cli/v9/configuring-npm/package-json)

andykb9b13 replied:
Cool. This is not something I've ever heard discussed up to this point but it makes a lot of sense. I'd be tearing my hair out. I'll check out the reading. Thanks!

mykltronn reviewed (commented):
A few general comments in addition to the inline comments.

First-- this is a pretty nice looking PR! It's a bit big, but I know front-end and React PRs specifically tend to be pretty length.

Second, I didn't run this locally, which I would normally do. For that reason, you should take some of what I say with a grain of salt. I'm hoping to run it locally sometime in the future ✅

andykb9b13 commented:
I've removed the front-end validation with yup resolver and am now relying on errors thrown from the resolver to send to the UI. 

- I don't think I've gotten to the root of the issue yet but this is as far as I can get right now I think. While I am not defining the shape of the data on the front-end AND the backend, I think I'm just defining it again but in a different spot (the resolver). I've added messages for validation directly in the model schema but I am having a really difficult time displaying those in the form on the front-end so it can be useful for the user. Basically I can't parse the error so it displays nicely. I've spent several hours going back and forth with it and I think I need assistance with that. At the moment, the validation is happening in the back-end which was the primary goal I think. 

- I've also tried to make the "try catch" blocks do less but I'm not sure if I've succeeded. I've got to trigger the login to happen with the data that is being created from the mutation so I've got to have those in the same scope somehow. I'm not sure how else to do it beyond how I've changed it here. 

- I've moved the styles into a separate file although they'll likely be turned into a larger theme and accessed through context at some point.

andykb9b13 commented on client/src/pages/Login.jsx:
```diff
@@ -5,58 +5,39 @@ import { LOGIN } from "../utils/mutations";
-  },
-};
+import { styles } from "../styles/cardstyles";
 
```
I figured out how to do the comments in the code :)

I've moved the individual styles to a separate file. For the time being, there are going to be individual style files for each page but eventually they will be consolidated into a theme and accessed through context.

andykb9b13 commented on client/src/pages/Login.jsx:
```diff
@@ -5,58 +5,39 @@ import { LOGIN } from "../utils/mutations";
+    Auth.login(data);
+    alert("Successfully logged in!");
+  };
 
```
I'm trying to make the `try` `catch` blocks "do" less so I've made a separate function to handle the login. I'm not sure if I'm just kicking the can down the road though.

andykb9b13 commented on client/src/pages/Login.jsx:
```diff
@@ -5,58 +5,39 @@ import { LOGIN } from "../utils/mutations";
+      {error && (
+        <FormHelperText sx={styles.errorText}>{error.message}</FormHelperText>
+      )}
       <form onSubmit={handleSubmit(onSubmit)}>
```
This is where I am displaying the error text being passed from the resolver.

andykb9b13 commented on server/schemas/resolvers.js:
```diff
@@ -70,6 +70,13 @@ const resolvers = {
+        throw new Error("Please enter your password");
+      }
+
       const user = teacher || student;
```
I've defined the errors to be show in the UI here in the resolver. Again, I'm not sure if this is missing the point but it's the closest I could get to it at the moment.
